File: python/html_to_org.py
import sys
import gumbo
import textwrap

from pathlib import PurePosixPath
from urllib.parse import urlparse, parse_qs, unquote


options = {"rewrap": True}



def extract_param_from_url(url: str, param: str) -> str:
    """Extract the page-id from the given URL."""
    parsed_url = urlparse(url)
    query_params = parse_qs(parsed_url.query)
    return query_params.get(param, [''])[0]


def extract_filename_from_url(url: str) -> str:
    path = urlparse(url).path
    return unquote(PurePosixPath(path).name)


def _traverse(node: gumbo.Node):
    # .findAll requires the .next pointer, which is what we're trying to add
    # when we call this, and so we manually supply a generator to yield the
    # nodes in DOM order.
    yield node
    try:
        for child in node.contents:
            for descendant in _traverse(child):
                yield descendant
    except AttributeError:
        # Not an element.
        return


def wrap(input: str,
         prefix: str,
         suffix=None,
         add_space: bool=False,
         remove_multispaces: bool=False):
    if suffix is None:
        suffix = prefix
    bprefix = prefix
    asuffix = suffix
    if add_space:
        bprefix = " " + prefix
        asuffix = suffix + " "
    if remove_multispaces:
        input = ' '.join(input.split()) # ignore spaces like all browsers do for most tags
    result = bprefix + input  + asuffix
    return result


def trimIt(string: str, leading=True, trailing=True):
    if(leading):
        while string.startswith(" ") or string.startswith(
                "\n") or string.startswith("\t"):
            string = string[1:]
    if(trailing):
        while string.endswith(" ") or string.endswith("\n") or string.endswith(
                "\t"):
            string = string[0:-1]
    return string

def fill(string, width=90):
    return textwrap.fill(" ".join(map(trimIt, string.split("\n"))), width)

def arrange_list_item(string, indent_prefix):
    string = fill(string)
    string = textwrap.indent(string, indent_prefix)
    string = trimIt(string, trailing=False)
    return string


def handle_tag(tag, result, list_indent_local, attributes=None, noFmt=[]):
    if tag in noFmt:
        return result
    if tag in ["H1", "H2", "H3", "H4"]:
        tag_s = tag.replace("H", "")
        result = ("*" * int(tag_s)) + " " + result + "\n"

    elif tag in ["TITLE"]:
        result = "\n#+TITLE: " + result + "\n"

    elif tag in ["STRONG", "B"]:
        result = wrap(result, "*", add_space=True, remove_multispaces=True)

    elif tag in ["TT"]:
        result = wrap(result, "=", add_space=True)

    elif tag in ["EM"]:
        result = wrap(result, "*", add_space=True)

    elif tag in ["I"]:
        result = wrap(result, "/", add_space=True)

    elif tag in ["DEL"]:
        result = wrap(result, "+", add_space=True)

    elif tag in ["SUB"]:
        result = wrap(result, "_{", "}", add_space=False)

    elif tag in ["SUP"]:
        result = wrap(result, "^{", "}", add_space=False)

    elif tag in ["P"]:
        if options["rewrap"]:
            result = fill(result)
        result = result + "\n\n"

    elif tag in ["BR"]:
        result = result + "\n\n"

    elif tag in ["TABLE"]:
        result = wrap(result, "\n")

    elif tag in ["CAPTION"]:
        result = wrap(result, "\n")

    elif tag in ["CODE", "PRE"]:
        language = ""
        if result.find("\n") == -1 and len(result) < 15:
            result = "=" + result + "="
        else:
            if(result.startswith(" =") and result.endswith("= ")):
                result = result[2:-2]
            result = "#+BEGIN_SRC {}\n".format(language) + result + "\n#+END_SRC\n\n"

    elif tag in ["TR"]:
        result = "| " + wrap(result, "", remove_multispaces=True) + "\n"

    elif tag in ["TH"]:
        result = result.replace('\n', '') + " | "

    elif tag in ["TD"]:
        result = result.replace('\n', '') + " | "

    elif tag in ["UL", "OL"]:
        result = result + "\n"

    elif tag in ["A"]:
        while result.startswith("["):
            result = result[1:]
        while result.endswith("]"):
            result = result[0:-1]
        if (attributes and "href" in attributes):
            if result == "":
                result = wrap(extract_param_from_url(attributes["href"], "page-id"), "[[", "]]", add_space=True) + '\n'
            else:
                result = wrap(extract_param_from_url(attributes["href"], "page-id") + "][" + result,
                              prefix="[[",
                              suffix="]]",
                              add_space=True) + '\n'

    elif tag in ["DL"]:
        result = result + "\n"

    elif tag in ["DT"]:
        result = (" " * list_indent_local) + "- " + result + " :: "

    elif tag in ["DD"]:
        result = arrange_list_item(result, (" " * list_indent_local) + "  ")
        result = result + "\n"

    elif tag in ["LI"]:
        result = arrange_list_item(result, (" " * list_indent_local) + "  ")
        result = (" " * list_indent_local) + "- " + result + "\n"

    elif tag in ["DIV", "SPAN"]:
        result = result

    elif tag in ["NAV", "META", "LINK", "SCRIPT", "STYLE"]:
        result = ""

    elif tag in ["IMG"]:
        result = wrap("file:./imgs/" + extract_filename_from_url(attributes["src"]), "[[", "]]")

    return result
    # else:

    # if tag in errors:
    #     errors[tag] = errors[tag] + 1
    # else:
    #     errors.add({tag: 0})


def skip_p(element: gumbo.Node):
    if hasattr(element, "tag"):
        if str(element.tag) == "DIV":
            if hasattr(element, "attributes"):
                attrs = {(attr.name).decode("utf-8"): (attr.value).decode("utf-8")
                         for attr in element.attributes}
                if "id" in attrs and attrs["id"] == "tree":
                    return True
    return False

def html_to_org_gumbo(element: gumbo.Node, list_indent=0, noFmt=[]):
    result = ""

    if hasattr(element, "tag"):
        if str(element.tag) in ["H1", "H2", "H3"]:
            list_indent = 0
        elif str(element.tag) in ["UL", "DL", "OL"]:
            list_indent += 3

    if hasattr(element, "type"):
        if str(element.type) == "TEXT":
            result += trimIt(element.contents.text.decode("utf-8"))
    if hasattr(element, "children"):
        for child in element.children:
            if skip_p(element): # my
                continue
            skip = []
            if hasattr(child, "tag") and hasattr(element, "tag"):
                if str(child.tag) == "CODE" and str(element.tag) == "PRE":
                    skip.append("CODE")
            result += html_to_org_gumbo(child, list_indent, noFmt=skip)

    if hasattr(element, "tag"):
        attrs = None
        if hasattr(element, "attributes"):
            attrs = {(attr.name).decode("utf-8"): (attr.value).decode("utf-8")
                     for attr in element.attributes}
        result = handle_tag(str(element.tag), result, list_indent, attrs, noFmt)

    return result


def dump_ast_gumbo(element: gumbo.Node, indent=0):
    sindent = (" " * indent)
    if hasattr(element, "type"):
        if str(element.type) == "TEXT":
            print(sindent + "With text -->" + element.contents.text.decode("utf-8"),
                  file=sys.stderr)

    if hasattr(element, "children"):
        for child in element.children:
            dump_ast_gumbo(child, indent + 2)

    if hasattr(element, "tag"):
        print(sindent + str(element.tag), file=sys.stderr)

# ------------- main

if (len(sys.argv) > 1):
    filename = sys.argv[1]
else:
    print("Usage: html_to_org HTMLFILENAME")
    quit(1)

with open(filename) as f:
    input = f.read()

with gumbo.gumboc.parse(input) as output:
    print(html_to_org_gumbo(output.contents.root.contents))

File: nim/html_to_org.nim
import gumbo
import algorithm
import parseopt
import tables
import sequtils
import httpclient
import os
import strutils

proc writeHelp()=
  echo getAppFilename() & " *.html"

proc writeVersion()=
  echo "v0.01"

var
  filename = ""
for kind, key, val in getopt():
  case kind
  of cmdArgument:
    filename = key
  of cmdLongOption, cmdShortOption:
    case key
      of "help", "h": writeHelp()
      of "version", "v": writeVersion()
      else:
        discard
  of cmdEnd: assert(false) # cannot happen
if filename == "":
  # no filename has been given, so we show the help:
  writeHelp()
  quit(1)

var is_http  = false
if filename.startsWith("http") or filename.startsWith("https"):
  is_http = true
type options_t = object
    full_rewrap: bool

var options : options_t = options_t(full_rewrap:false)

let show_debug = false
import times
import htmlparser
let n1 = epochtime()
if is_http == false:
  let nim_html = loadHtml(filename)  ## measurement purposes
let n2 = epochtime()

import htmlgen
let g1 = epochtime()
var html = ""
if is_http:
  var client = newHttpClient()
  html = client.getContent(filename)
else:
  html = readFile(filename)

var out1 = gumbo.gumbo_parse(html)
let g2 = epochtime()

# root
var root = out1.root[]

if show_debug:
  stderr.write "root: " & $gumbo_normalized_tagname( root.v.element.tag) &
    "with " & $root.v.element.children.length & " children"


proc tableify_attributes (gv:GumboVector): Table[string, string]=
  result = initTable[string,string]()

  let GRchildrenVector = gv
  for i in countUp[int](a=0,b=int(GRchildrenVector.length)-1):
    let test = GRchildrenVector.data
    var nodeP = cast[ptr ptr GumboAttribute](cast[int](test) +% i* sizeof(ptr pointer) ) # pointer arithmagic
    var node = nodeP[][]
    let name = $node.name
    let value = $node.value
    result.add(name, value)

proc tag_toStr(tag: GumboTag): string=
  result = ($tag).replace("GUMBO_TAG_", "")


proc trimIt(input:string, leading:bool=true, trailing:bool=true):string {.procvar.}=
  result = input
  if result in ["\n"]:
    result = ""
  while (result.startsWith(" ") or result.startsWith("\n") or result.startsWith("\t")):
    result = result[1..result.len-1]
  while (result.endsWith(" ") or result.endsWith("\n") or result.endsWith("\t")):
    result = result[0..result.len - 2]

var unknown = initTable[string, int]()

#import templates
template foreach_child(node:untyped,children:GumboVector, body:untyped):untyped=
  for i in countUp[int](a=0,b=int(children.length)-1):
    let test = children.data
    let nodeP = cast[ptr ptr GumboNode](cast[int](test) +% i* sizeof(ptr pointer) ) # pointer arithmagic
    node = nodeP[][]
    body

proc print_ngumbo_to_org*(gumbon: GumboNode, list_indent:int=0, noFmt:openArray[string]=[]):string=
  result = ""
  let gelem = gumbon.v.element
  let attributes = tableify_attributes(gelem.attributes)
  let tagStr = tag_toStr(gelem.tag)
  var list_indent_local = list_indent
  if (gelem.tag in {GUMBO_TAG_H1, GUMBO_TAG_H2, GUMBO_TAG_H3}):
    list_indent_local = 3

  if (gelem.tag in {GUMBO_TAG_UL, GUMBO_TAG_DL, GUMBO_TAG_OL}):
    list_indent_local = list_indent + 3
  var skip :seq[string] = @[]

  var child:GumboNode
  foreach_child(child,gelem.children):
    if(child.`type` == GUMBO_NODE_ELEMENT):
      if tagStr == "CODE" and tag_toStr(gelem.tag) == "PRE":
        skip.add("PRE")
      if tagStr == "PRE" and tag_toStr(gelem.tag) == "CODE":
        skip.add("CODE")
      result &= print_ngumbo_to_org(child, list_indent_local, skip)
    if(child.`type` == GUMBO_NODE_TEXT):
      result &= ($child.v.text.text).trimIt
    if(child.`type` == GUMBO_NODE_DOCUMENT):
      stderr.write "no impl"


  if defined(result) and result.trimIt == "" :
     return result

  proc wrap(input:string, prefix,suffix: string, add_space:bool=false, add_newlines:bool=false ):string=
    var bprefix = prefix
    var asuffix = suffix
    if add_space:
      bprefix = " " & prefix
      asuffix = suffix & " "

    result = bprefix & input & asuffix

  proc wrap(input:string, wrapStr: string, add_space:bool=false, add_newlines:bool=false ):string=
    wrap(input, wrapStr, wrapStr, add_space, add_newlines)

  proc fill(s:string, width:int=90):string=
    result = wordWrap(s.split("\n").mapIt(trimIt(it)).join(" "), maxLineWidth = width, splitLongWords = false)

  proc arrange_list_item(s:string, indent_prefix:string):string=
    result = fill(s)
    result = indent(result, len(indent_prefix))
    result = trimIt(result, trailing=false)

  if(tagStr in noFmt):
    return result

  case tagStr:
    of "H1" , "H2" , "H3", "H4", "H5", "H6":
      let tag_s = tagStr.replace("H", "")
      let level = parseInt($tag_s[0])
      result = repeat("*", level) & " " & result & "\n"
      if attributes.contains("id"):
        stderr.write tagStr & ": " & attributes["id"] & "\n"

    of "TITLE":
      result = "\n#+TITLE: " & result & "\n"

    of "STRONG", "B":
      result = wrap(result, "*", true)

    of "TT":
      result = wrap(result, "=", true)

    of "EM":
      result = wrap(result, "*", true)

    of "I":
      result = wrap(result, "/", true)

    of "DEL":
      result = wrap(result, "+", true)

    of "SUB":
      result = wrap(result, "_{","}", false)

    of "SUP":
      result = wrap(result, "^{","}", false)

    of "P":
      if options.full_rewrap:
        result = wordWrap(result.split("\n").mapIt(trimIt(it)).join(" "), splitLongWords = false)
      result = result & "\n\n"

    of "BR":
      result = result & "\n\n"

    of "TABLE":
      result = wrap(result, "\n")

    of "CODE", "PRE":
      if result.find("\n") == -1 and result.len < 15:
        result = "=" & result & "="
      else:
        if(result.startsWith(" =") and result.endSwith("= ")):
          result = result[2..result.len-3]
        result = "#+BEGIN_SRC\n" & result & "\n#+END_SRC\n\n"

    of "TR":
      result &= "\n"

    of "TD":
      result = " | " & result & " | "

    of "UL", "OL":
      result = result & "\n"

    of "A":
      while result.startsWith("["):
        result = result[1..result.len-1]
      while result.endsWith("]"):
        result = result[0..result.len-2]
      if  attributes.contains("href" ):
        if result == "":
          result = " [[" & attributes["href"].trimIt & "]] "
        else:
          result = (attributes["href"].trimIt & "][" & result).wrap("[[","]]", add_space = true)

    of "DL":
      result =  result & "\n"

    of "DT":
      result = repeat(" ", list_indent_local) & "- " & result & " :: "

    of "DD":
      result = arrange_list_item(result, repeat(" ", list_indent_local) & "  ")
      result = result & "\n"

    of "LI":
      result = arrange_list_item(result, repeat(" ", list_indent_local) & "  ")
      result = repeat(" ", list_indent_local) & "- " &  result & "\n"

    of "DIV", "SPAN":
      discard

    of "NAV", "META", "LINK", "SCRIPT", "STYLE":
      result = ""
    else:
      if unknown.contains($gelem.tag):
        unknown[$(gelem.tag)] = unknown[$gelem.tag]+1
      else:
        unknown.add($gelem.tag, 0)


echo print_ngumbo_to_org(root)

if show_debug:
  let g3 = epochtime()

  stderr.write "\nunknown key type:\n"
  for key in unknown.keys:
    stderr.write "  - " & key & "\n"

  stderr.write "\nTiming: \n"
  stderr.write "  - gumbo:" & $(g2-g1) & " + " & $(g3-g2) & "\n"
  stderr.write "  - nim:" & $(n2-n1) & "\n"


proc dump_ast*(gumbon: GumboNode, indent:int=0)=
  let gelem = gumbon.v.element
  let ide = repeat(" ", indent)
  stderr.write ide & " " & tag_toStr(gelem.tag) & "\n"
  var child:GumboNode
  foreach_child(child,gelem.children):
    if(child.`type` == GUMBO_NODE_ELEMENT):
      dump_ast(child, indent + 2 )
    if(child.`type` == GUMBO_NODE_TEXT):
      stderr.write ide & "  " & "with text --> " & escape(($child.v.text.text).trimIt) & "\n"
    if(child.`type` == GUMBO_NODE_DOCUMENT):
      stderr.write "no impl"

if show_debug:
  dump_ast(root)


# Local Variables:
# firestarter: "nim -d:release -d:ssl c %f || notify-send -u low 'nim' 'compil%f'"
# End:
